ydb/core/cms/console/console_handshake.cpp

```diff
@@ -0,0 +1,104 @@
+        if (Self.CurrentInterconnectSession) {
+            endSessionEvent->Rewrite(TEvInterconnect::EvForward, Self.CurrentInterconnectSession);
+            ctx.Send(endSessionEvent.release());
+        }
+        else {
+            return;
+        }
+    }
+    Self.CurrentSenderId = ev->Sender;
+    Self.CurrentPipeServerId = ev->Recipient;
+    Self.CurrentInterconnectSession = ev->InterconnectSession;
+    auto response = std::make_unique<TEvBlobStorage::TEvControllerProposeConfigResponse>();
+    auto& responseRecord = response->Record;
+
+    if (YamlVersion > record.GetConfigVersion()) {
+        responseRecord.SetStatus(NKikimrBlobStorage::TEvControllerProposeConfigResponse::UnexpectedConfig);
+        responseRecord.SetProposedConfigVersion(proposedConfigVersion);
+        responseRecord.SetConsoleConfigVersion(YamlVersion);
+        SendInReply(ev->Sender, Self.CurrentInterconnectSession, std::move(response));
+        LOG_ALERT_S(ctx, NKikimrServices::CMS, "Unexpected proposed config.");
+        return;
+    }
+    if (YamlVersion == record.GetConfigVersion()) {
+        responseRecord.SetStatus(NKikimrBlobStorage::TEvControllerProposeConfigResponse::CommitIsNotNeeded);
+        SendInReply(ev->Sender, Self.CurrentInterconnectSession, std::move(response));
+        return;
+    }
+    if (proposedConfigHash != currentConfigHash) {
+        responseRecord.SetStatus(NKikimrBlobStorage::TEvControllerProposeConfigResponse::HashMismatch);
+        responseRecord.SetProposedConfigHash(proposedConfigHash);
+        responseRecord.SetConsoleConfigHash(currentConfigHash);
+        SendInReply(ev->Sender, Self.CurrentInterconnectSession, std::move(response));
+        LOG_ALERT_S(ctx, NKikimrServices::CMS, "Config hash mismatch.");
+        return;
+    }
+    responseRecord.SetStatus(NKikimrBlobStorage::TEvControllerProposeConfigResponse::CommitIsNeeded);
+    SendInReply(ev->Sender, Self.CurrentInterconnectSession, std::move(response));
+}
+
+void TConfigsManager::Handle(TEvBlobStorage::TEvControllerConsoleCommitRequest::TPtr &ev, const TActorContext &ctx) {
+    auto response = std::make_unique<TEvBlobStorage::TEvControllerConsoleCommitResponse>();
+    auto& record = response->Record;
+    const auto& yamlConfig = ev->Get()->Record.GetYAML();
+
+    if (Self.CurrentSenderId != ev->Sender || Self.CurrentPipeServerId != ev->Recipient) {
+        if (Self.CurrentSenderId == ev->Sender) {
+            return;
+        }
+        record.SetStatus(NKikimrBlobStorage::TEvControllerConsoleCommitResponse::SessionMismatch);
+        SendInReply(ev->Sender, Self.CurrentInterconnectSession, std::move(response));
+        return;
+    }
+
+    auto replaceConfigEv = TEvConsole::TEvReplaceYamlConfigRequest();
+    replaceConfigEv.Record.MutableRequest()->set_config(yamlConfig);
+    auto handle = new IEventHandle(ev->Sender, Self.SelfId(), &replaceConfigEv, 0, ev->Cookie);
```

Comment:
again, recipient goes first
and you can't send local event, it is expected to be dynamically allocated and ownership is taken by the actor system as you invoke Send